qadence_protocols/mitigations/readout.py

```diff
@@ -116,48 +192,42 @@ def mitigation_minimization(
     corrected_counters: list[Counter] = []
 
     for sample in samples:
-        bitstring_length = 2**n_qubits
-        # List of bitstrings in lexicographical order.
-        ordered_bitstrings = [f"{i:0{n_qubits}b}" for i in range(bitstring_length)]
+        ordered_bitstrings = [bin(k)[2:].zfill(n_qubits) for k in range(2**n_qubits)]
         # Array of raw probabilites.
         p_raw = np.array([sample[bs] for bs in ordered_bitstrings]) / n_shots
 
         if optimization_type == ReadOutOptimization.CONSTRAINED:
-            # Initial random guess in [0,1].
-            p_corr0 = np.random.rand(bitstring_length)
-            # Stochasticity constraints.
-            normality_constraint = LinearConstraint(
-                np.ones(bitstring_length).astype(int), lb=1.0, ub=1.0
-            )
-            positivity_constraint = LinearConstraint(
-                np.eye(bitstring_length).astype(int), lb=0.0, ub=1.0
-            )
-            constraints = [normality_constraint, positivity_constraint]
-            # Minimize the corrected probabilities.
-            res = minimize(
-                corrected_probas, p_corr0, args=(noise_matrices, p_raw), constraints=constraints
-            )
-            p_corr = res.x
+            p_corr = constrained_inversion(noise_matrices, p_raw)
 
         elif optimization_type == ReadOutOptimization.MLE:
             noise_matrices_inv = list(map(matrix_inv, noise_matrices))
             # Compute corrected inverse using matrix inversion and run MLE.
             p_corr = mle_solve(tensor_rank_mult(noise_matrices_inv, p_raw))
+
+        elif optimization_type == ReadOutOptimization.MTHREE:
+            confusion_matrix_subspace = normalized_subspace_kron(noise_matrices, p_raw.nonzero()[0])
+            # GMRES (Generalized minimal residual) for linear equations in higher dimension
+            p_corr, exit_code = gmres(confusion_matrix_subspace, p_raw)
+
+            if exit_code != 0:
+                logger.warning(f"gmres did not converge, exited with code {exit_code}")
```

Comment:
```suggestion
                logger.warning(f"GMRES did not converge, exited with code {exit_code}")
```

Also: do you want to raise an exception here ?

Reply:
I think this should be sufficient, I cant think of use case now in this context, can be added if required later